alf/algorithms/dynamics_learning_algorithm.py

```diff
@@ -253,3 +304,98 @@ def train_step(self, time_step: TimeStep, state: DynamicsState):
         state = state._replace(feature=feature)
 
         return AlgStep(output=(), state=state, info=info)
+
+
+@gin.configurable
+class StochasticDynamicsAlgorithm(DeterministicDynamicsAlgorithm):
+    """Stochastic Dynamics Learning Module
+
+    This module learns the dynamics of environment with a
+    stochastic model.
+    """
+
+    def __init__(self,
+                 action_spec,
+                 feature_spec,
+                 hidden_size=256,
+                 num_replicas=1,
+                 dynamics_network: DynamicsNetwork = None,
+                 name="StochasticDynamicsAlgorithm"):
+        """Create a StochasticDynamicsAlgorithm.
+
+        Args:
+            hidden_size (int|tuple): size of hidden layer(s)
+            num_replicas (int): number of network replicas to be used
+                in the ensemble for dynamics learning
+            dynamics_network (Network): network for predicting next feature
+                based on the previous feature and action. It should accept
+                input with spec [feature_spec, encoded_action_spec] and output
+                a tensor of shape feature_spec. For discrete action,
+                encoded_action is an one-hot representation of the action.
+                For continuous action, encoded action is the original action.
+        """
+
+        assert dynamics_network._prob, "should use probabistic network"
+        super().__init__(
+            action_spec=action_spec,
+            feature_spec=feature_spec,
+            hidden_size=hidden_size,
+            num_replicas=num_replicas,
+            dynamics_network=dynamics_network)
+
+    def predict_step(self, time_step: TimeStep, state: DynamicsState):
+        """Predict the next observation given the current time_step.
+                The next step is predicted using the prev_action from time_step
+                and the feature from state.
+        """
+        action = self._encode_action(time_step.prev_action)
+        obs = state.feature
+
+        # perform preprocessing
+        observations = self._expand_to_replica(obs, self._feature_spec)
+        actions = self._expand_to_replica(action, self._action_spec)
+
+        dist, network_states = self._dynamics_network((observations, actions),
+                                                      state=state.network)
+
+        forward_deltas = dist.sample()
+
+        forward_preds = observations + forward_deltas
+        state = state._replace(feature=forward_preds, network=network_states)
+        return AlgStep(output=forward_preds, state=state, info=dist)
```

Comment:
Seems that `info` has no use here.

Reply:
It is used in train_step for doing loss calculation using ```dist.log_prob(observation)```.
Now moved ```dist``` into ```DynamicsInfo``` for consistency.